KEP-79: Hierarchical Cohorts (#1531)
kubernetes-sigs/kueue: keps/79-hierarchical-cohorts/README.md

# KEP-79: Hierarchical Cohorts

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
      - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit Tests](#unit-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

Introduce Cohort top-level object to allow setting multi-level quota 
hierarchy, with advanced borrowing, lending mechanisms.

## Motivation

The current 2-level hierarchy (ClusterQueues and Cohorts) is not expressive

> enough to handle complex use ceases of large organizations with tree-like

Reviewer: ```suggestion
enough to handle complex use cases of large organizations with tree-like
```

Author: Done.

> enough to handle complex use ceases of large organizations with tree-like

Author: Done.

team and quota/budget structures.

### Goals

* Create a multi-level hierarchy for advanced quota management. 
* Be compatible with the existing ClusterQueue API and mechanics.
* Allow setting constraints about borrowing and lending at all levels.
* Provide quota for groups of Queues.

### Non-Goals

* Change the existing API and mechanics in a not backward-compatible way.

> * Introduce a new, separate API.

Author: Done.

* Introduce new ways of fair sharing, like ratio-based sharing (at least 
not in this KEP).
* Introduce additional preemption models (this will be in a separate KEP).

## Proposal

> Introduce a new object called Cohort with the simila quota provisioning mechanism

Reviewer: ```suggestion
Introduce a new object called Cohort with the similar quota provisioning mechanism
```

Author: Done.

> Introduce a new object called Cohort with the simila quota provisioning mechanism

Author: Done.

as ClusterQueue. Chorots additionally may specify its parent, another Cohort,
forming together a tree-like organization structure. ClusterQueues will still be able to specify
the Cohort they belong to. The Cohort mentioned by ClusterQueue doesn't require
an actual object to be present. If such is not provided, it is understood 
that the Cohort doesn't provide any quota, has no parent, doesn't belong to any bigger
structure or has any non-default settings.

The difference between ClusterQuota and Cohort will be that:

* ClusterQueues are leaves in the organization tree, Cohorts are inner nodes. 
* Cohort doesn't accept any workloads.

> * Quota provided at the Cohort level is to be shared with the entire organization and doesn't 

Author: Done.

>   have a owning ClusterQueue.

Reviewer: ```suggestion
  have an owning ClusterQueue.
```

Author: Done.

workloads. Both are given some quota that is further distributed among the subteams. I want to grant 
the production workloads the ability to borrow research quota if needed, but not the other way round.

#### Story 2

I have a couple organizations that have dedicated resources. The organizations should not borrow 
from each other, however I want to have an additional "special" queue, with low priority jobs, that can 
borrow unused capacity from any of the organizations.

### Risks and Mitigations

* Users may not create a Cohort object for Cohorts mentioned in ClusterQueues - this is fine,
we will assume that the Cohort doesn't have any quota or other constraints/settings. 

* Users may provide non-existent parent information for a Cohort - this is fine, 
we will simply assume that the parent has no quota or other constraints/settings.

* Users may create a cycle in the Cohort hierarchy - Kueue will stop all new admissions within
the entire tree. The already admitted workloads will be allowed to continue. Appropriate 
ClusterQueue/Cohort Status Conditions will be set and Events emited.

> * Scheduling and preemption may require mor computation/resources.

Reviewer: ```suggestion
* Scheduling and preemption may require more computation/resources.
```

Author: Done

## Design Details

The Cohort API will initially start only with the basic functionality. Additional policies
regarding sharing Cohort resources can be added later.

```go

type Cohort struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec   CohortSpec   `json:"spec,omitempty"`
    Status CohortStatus `json:"status,omitempty"`
}

> type CohortSpec struct {

Reviewer: How does the Cohort API specify the parent cohort? I suppose we need another field in spec.

Author: Good catch, somehow got lost during copy-paste. Added.

workloads. Both are given some quota that is further distributed among the subteams. I want to grant 
the production workloads the ability to borrow research quota if needed, but not the other way round.

#### Story 2

I have a couple organizations that have dedicated resources. The organizations should not borrow 
from each other, however I want to have an additional "special" queue, with low priority jobs, that can 
borrow unused capacity from any of the organizations.

### Risks and Mitigations

* Users may not create a Cohort object for Cohorts mentioned in ClusterQueues - this is fine,
we will assume that the Cohort doesn't have any quota or other constraints/settings. 

* Users may provide non-existent parent information for a Cohort - this is fine, 
we will simply assume that the parent has no quota or other constraints/settings.

* Users may create a cycle in the Cohort hierarchy - Kueue will stop all new admissions within
the entire tree. The already admitted workloads will be allowed to continue. Appropriate 
ClusterQueue/Cohort Status Conditions will be set and Events emited.

* Scheduling and preemption may require more computation/resources.

> ## Design Details

Reviewer: If we have `ClusterQueue.preemption.reclaimWithinCohort=Any`, does it mean preemption is allowed within the first level of cohort, or the whole hierarchy?

Author: Whole hierarchy.

Reviewer: I would suggest adding such a note on preemption in KEP.

Author: Added to proposal section.

The Cohort API will initially start only with the basic functionality. Additional policies
regarding sharing Cohort resources can be added later.

```go

type Cohort struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec   CohortSpec   `json:"spec,omitempty"`
    Status CohortStatus `json:"status,omitempty"`
}

type CohortSpec struct {
    // Cohort parent name. The parent Cohort object doesn't have to exist.
    // In such case, it is assumed that parent simply doesn't have any
    // quota and limits and doesn't have any other custom settings.

>     Parent *string `json:"resourceGroups,omitempty"`

Reviewer: ```suggestion
    Parent *string `json:"parent,omitempty"`
```

Author: Done.

* `T(x, r)` - returns the amount of resource r that is provided at the level of x from ClusterQueues
and Cohorts that are either x or children of x (possibly indirect). In other words, how much of resource r can come from 
the subtree. The value may be negative, what means that the subtree is borrowing from the outside of the subtree (the rest 
of the hierarchy)

* `D(x,r)` - returns the amount of resource r that is provided at the level of x from outside of the subtree. The value may
be negative, what means that the rest of the hierarchy borrows from the subtree. 

Obviously, with the correct admission process, for any x and r, $$T(x,r) + D(x,r)  \geq 0$$
Otherwise there would be an unbalanced debt at level x. If there is no debt at any level then the admission is correct.

New workload can be admitted to a ClusterQueue if the above conditions stays true at all elements of the hierarchy.
Checking it only for the ClusterQueue may not be enough as changes related to borrowing may hit limits elsewhere.

`T(x,r)` and `D(x,r)` can be relatively easily calculated while traversing the Cohort tree.

* `T(x,r)` when x is a ClusterQueue:

> $$T(x,r) = quota(x,r) - usage(x,r)$$

Reviewer: I'm not sure if these formulas are "definitions" or "corollaries" of some other assumptions. Please make it clear is the relationship between the formulas and the statement: "T(x,r) + D(x,r)  >= 0".

Author: After some consideration i came to conclusion that we can get rid of D(x,r) in favor of just borrowingLimit(x,r). I pivoted the description there a bit as well. PTAL.

Reviewer: Thanks, it reads much better now. I see that the formulas are definitions, and in this section we argue that verifying the condition `T(x,r) + borrowingLimit(x,r) >= 0` at all levels is enough to check the correctness of the admission.

One little puzzling thing is that the definition of `D(x,r)` was different than `borrowingLimit(x,r)`, so it seems the condition actually changed during the re-write. I assume the current version is correct. I will open another dedicated comment if I have some doubts, but I consider the "structural" comment addressed.

Author: `D(x,r)` was in fact borrowing limit trimmed to what the rest of the tree can provide. However, we check the provisioning at the level above, so there is no need for extra trimming. It doesn't provide any more information.

enough to handle complex use cases of large organizations with tree-like
team and quota/budget structures.

### Goals

* Create a multi-level hierarchy for advanced quota management. 
* Be compatible with the existing ClusterQueue API and mechanics.
* Allow setting constraints about borrowing and lending at all levels.
* Provide quota for groups of Queues.

### Non-Goals

* Change the existing API and mechanics in a not backward-compatible way.
* Introduce an alternative API to ClusterQueue.
* Introduce new ways of fair sharing, like ratio-based sharing (at least 
not in this KEP).
* Introduce additional preemption models (this will be in a separate KEP).

## Proposal

Introduce a new object called Cohort with the similar quota provisioning mechanism
as ClusterQueue. Chorots additionally may specify its parent, another Cohort,
forming together a tree-like organization structure. ClusterQueues will still be able to specify
the Cohort they belong to. The Cohort mentioned by ClusterQueue doesn't require
an actual object to be present. If such is not provided, it is understood 
that the Cohort doesn't provide any quota, has no parent, doesn't belong to any bigger

> structure or has any non-default settings.

Author: I'm rather against automatic object creation.

> The difference between ClusterQuota and Cohort will be that:

Author: Done.

* Lending limit specified at the Cohort levels means that the rest of the organization tree
  cannot borrow more from the subtree than the given value.

Preemmptions and resoruce reclamation will happen among the whole cohort structure,
in the similar fashion as they are executed now.

### User Stories (Optional)

#### Story 1

I have two multi-team organizations in the company. One that does research and one that runs production
workloads. Both are given some quota that is further distributed among the subteams. I want to grant 
the production workloads the ability to borrow research quota if needed, but not the other way round.

#### Story 2

I have a couple organizations that have dedicated resources. The organizations should not borrow 
from each other, however I want to have an additional "special" queue, with low priority jobs, that can 
borrow unused capacity from any of the organizations.

### Risks and Mitigations

* Users may create a cycle in the Cohort hierarchy - Kueue will stop all new admissions within
the entire tree. The already admitted workloads will be allowed to continue. Appropriate 
ClusterQueue/Cohort Status Conditions will be set and Events emited.

> * Scheduling and preemption may require more computation/resources.

Author: I kind-of trust in the common sense of the user. If they create 100 cohorts in a tree structure, so be it. We don't put limits on the number of pods in K8S cluster, or jobs or anything like that.

This is just a note that, with the complex structure, scheduling and preemption will be a bit more expensive. Still, i suspect that the impact will be marginal and the biggest bottleneck will be etcd communication anyway.

regarding sharing Cohort resources can be added later.

```go

type Cohort struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec   CohortSpec   `json:"spec,omitempty"`
    Status CohortStatus `json:"status,omitempty"`
}

type CohortSpec struct {
    // Cohort parent name. The parent Cohort object doesn't have to exist.
    // In such case, it is assumed that parent simply doesn't have any
    // quota and limits and doesn't have any other custom settings.
    Parent *string `json:"parent,omitempty"`

    // resourceGroups describes groups of resources that the Cohort can
    // share with ClusterQueues within the same group of Cohorts/ClusterQueues.
    // Each resource group defines the list of resources and a list of flavors
    // that provide quotas for these resources.
    // Each resource and each flavor can only form part of one resource group.
    // resourceGroups can be up to 16.
    //
    // BorrowingLimit specifies how much ClusterQueues under this Cohort can borrow
    // from ClusterQueues/Cohorts that are NOT under this Cohort. For Cohorts without

>     // a parent(top of the hierarchy) the BorrowingLimit has to be 0.

Reviewer 2: ```suggestion
    // a parent (top of the hierarchy) the BorrowingLimit has to be 0.
```

Author: Done.

    //
    // LendingLimit specifies how much ClusterQueues that are NOT under this Cohort
    // can borrow from the ClusterQueues/Cohorts that are under this Cohort.
    // If any of the Limits is not specified it means that there is no limit
    // and ClusterQueues can borrow/lend as much as they want/have.
    // 
    // +listType=atomic
    // +kubebuilder:validation:MaxItems=16

>     ResourceGroups []ResourceGroup `json:"resourceGroups,omitempty"`

Reviewer 2: Ah, this means existing `ResourceGroup`, right?

https://github.com/kubernetes-sigs/kueue/blob/1579d75e5a35bdb4ab24c64b2da4e4effa98bbc8/apis/kueue/v1beta1/clusterqueue_types.go#L137

>     ResourceGroups []ResourceGroup `json:"resourceGroups,omitempty"`

Author: yes, existing ResourceGroup.

}

type CohortStatus struct {
    // conditions hold the latest available observations of the Conditions
    // current state.
    // +optional
    // +listType=map
    // +listMapKey=type
    // +patchStrategy=merge
    // +patchMergeKey=type

>     Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`

Author: Added.

  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
      - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit Tests](#unit-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

Introduce Cohort top-level object to allow setting multi-level quota 
hierarchy, with advanced borrowing, lending mechanisms.

## Motivation

The current 2-level hierarchy (ClusterQueues and Cohorts) is not expressive
enough to handle complex use cases of large organizations with tree-like
team and quota/budget structures.

### Goals

* Create a multi-level hierarchy for advanced quota management. 
* Be compatible with the existing ClusterQueue API and mechanics.
* Allow setting constraints about borrowing and lending at all levels.
* Provide quota for groups of Queues.

### Non-Goals

* Change the existing API and mechanics in a not backward-compatible way.
* Introduce a new, separate API.
* Introduce new ways of fair sharing, like ratio-based sharing (at least 
not in this KEP).

> * Introduce additional preemption models (this will be in a separate KEP).

Author: We have some ideas for new preemption modes, but they are not formalized into an issue yet.

## Proposal

Introduce a new object called Cohort with the similar quota provisioning mechanism

> as ClusterQueue. Chorots additionally may specify its parent, another Cohort,

Author: Done.

forming together a tree-like organization structure. ClusterQueues will still be able to specify
the Cohort they belong to. The Cohort mentioned by ClusterQueue doesn't require
an actual object to be present. If such is not provided, it is understood 
that the Cohort doesn't provide any quota, has no parent, doesn't belong to any bigger
structure or has any non-default settings.

The difference between ClusterQuota and Cohort will be that:

* ClusterQueues are leaves in the organization tree, Cohorts are inner nodes. 
* Cohort doesn't accept any workloads.
* Quota provided at the Cohort level is to be shared with the entire organization and doesn't 
  have an owning ClusterQueue.
* Borrowing limit specified at the Cohort level means that the entire subtree cannot borrow more 
  from the rest of the organization tree than the given value.
* Lending limit specified at the Cohort levels means that the rest of the organization tree
  cannot borrow more from the subtree than the given value.

> ### User Stories (Optional)

Reviewer 3: worth putting some yaml snippets to gather feedback.

Author: Added descriptions how the proposal handles user stories. Yamls would be unnecessarily long.

#### Story 1

I have two multi-team organizations in the company. One that does research and one that runs production
workloads. Both are given some quota that is further distributed among the subteams. I want to grant 
the production workloads the ability to borrow research quota if needed, but not the other way round.

#### Story 2

I have a couple organizations that have dedicated resources. The organizations should not borrow 
from each other, however I want to have an additional "special" queue, with low priority jobs, that can 
borrow unused capacity from any of the organizations.

### Risks and Mitigations

* Users may not create a Cohort object for Cohorts mentioned in ClusterQueues - this is fine,
we will assume that the Cohort doesn't have any quota or other constraints/settings. 

* Users may provide non-existent parent information for a Cohort - this is fine, 

> we will simply assume that the parent has no quota or other constraints/settings.

Reviewer 3: these 2 are Notes/caveats, not Risks.

Author: Removed.

* Users may create a cycle in the Cohort hierarchy - Kueue will stop all new admissions within
the entire tree. The already admitted workloads will be allowed to continue. Appropriate 
ClusterQueue/Cohort Status Conditions will be set and Events emited.

> * Scheduling and preemption may require more computation/resources.

Reviewer 3: Any mitigation? Or just a heads up in the documentation?

```go

type Cohort struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec   CohortSpec   `json:"spec,omitempty"`
    Status CohortStatus `json:"status,omitempty"`
}

type CohortSpec struct {
    // Cohort parent name. The parent Cohort object doesn't have to exist.
    // In such case, it is assumed that parent simply doesn't have any
    // quota and limits and doesn't have any other custom settings.
    Parent *string `json:"parent,omitempty"`

    // resourceGroups describes groups of resources that the Cohort can
    // share with ClusterQueues within the same group of Cohorts/ClusterQueues.
    // Each resource group defines the list of resources and a list of flavors
    // that provide quotas for these resources.
    // Each resource and each flavor can only form part of one resource group.
    // resourceGroups can be up to 16.
    //
    // BorrowingLimit specifies how much ClusterQueues under this Cohort can borrow
    // from ClusterQueues/Cohorts that are NOT under this Cohort. For Cohorts without
    // a parent(top of the hierarchy) the BorrowingLimit has to be 0.
    //

>     // LendingLimit specifies how much ClusterQueues that are NOT under this Cohort

Reviewer 3: FYI @kerthcet

    // +patchMergeKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`

    // Additional stats may be added in the future, like the number 
    // of admitted workloads, their usage etc, based on the user feedback.
}
```

Currently, with 2-level hierarchy for each Cohort and ClusterQueue, Kueue
makes sure that the following balances are kept:

* ClusterQueues don't use more resources than they have and could possibly borrow.
* Within a Cohort, the total amount of requested capacity doesn't exceed the total quota
from all ClusterQueues, constrained by LendingLimit.

Admission of a new workload can happen if both balances are kept after 
adding the new workload. Kueue doesn't track who is borrowing/lending from who. It
is enough that balances are kept and with good balances, there exists such borrower-lending 
mapping that fulfills all needs. 

With Hierarchical Cohorts, Kueue will be checking the whole Cohort subtree whether the
correct balances are kept. To be more precise what it means
let's define a function `T(x,r)` that takes either ClusterQueue x 
or Cohort x and resource r (from a specific resource flavor). 

> `T(x, r)` returns the amount of resource r that is provided at the level of x from ClusterQueues

Reviewer 3: ```suggestion
`T(x, r)` returns the amount of resource r that is available at the level of x from ClusterQueues
```

Author: Done.

and Cohorts that are either x or children of x (possibly indirect). In other words, how much of resource r can come from 
the subtree. The value may be negative, what means that the subtree is borrowing from the outside of the subtree (the rest 
of the hierarchy)

`T(x,r)` can be relatively easily calculated while traversing the Cohort tree.

* `T(x,r)` when x is a ClusterQueue:
$$T(x,r) = quota(x,r) - usage(x,r)$$

* `T(x,r)` when x is a Cohort:
$$T(x,r) = quota(x,r) + \sum_{c \in children(x)} min(lendingLimit(c,r), T(c,r))$$

Obviously, with the correct admission process, for any x and r, `T(x,r)  >=  -borrowingLimit(x,r)`
Otherwise there would be too big debt at level x - some subtree is requesting more than allowed.

Slightly less obvious, but also true is: **If there is no too big debt at any level then the admission is correct**.

> Negative `T(x,r)` presents the total amount of resources that a subtree needs. Positive `T(x,r)` presents the total 

Reviewer 3: ```suggestion
Negative `T(x,r)` presents the total amount of resources that a subtree is borrowing. Positive `T(x,r)` presents the total 
```

Author: Done.

"passed" to the needing subtrees. Then, after this passing, the previously negative subtree becomes positive, and 
we can re-apply the logic there. All negative sub-sub-trees can be balanced out by positive subtrees and the capacity
that coming from "above". And so on and so on, up to reaching individual ClusterQueues.

So a new workload can be admitted to a ClusterQueue if and only if, after admission, `T(x,r)  >=  -borrowingLimit(x,r)` 
stays true at all elements of the hierarchy.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

#### Unit Tests

As the hierarchical cohorts reside entirely inside Kueue, most of the 
tests will be done as unit and integration tests, checking things like:

* Existing functionality at 2-levels.
* Long-distance borrowing on multi-level hierarchy.
* Lending/borrowing limits placed on many levels.
* Preemptions across hierarchy.

### Graduation Criteria

This is an a core API element and will graduate 

## Implementation History

> * 2023.12.29 - Initial KEP 

Reviewer 3: ```suggestion
* 2023.12.29 - KEP: API and semantics
```

Author: Done.

    Parent *string `json:"parent,omitempty"`

    // resourceGroups describes groups of resources that the Cohort can
    // share with ClusterQueues within the same group of Cohorts/ClusterQueues.
    // Each resource group defines the list of resources and a list of flavors
    // that provide quotas for these resources.
    // Each resource and each flavor can only form part of one resource group.
    // resourceGroups can be up to 16.
    //
    // BorrowingLimit specifies how much ClusterQueues under this Cohort can borrow
    // from ClusterQueues/Cohorts that are NOT under this Cohort. For Cohorts without
    // a parent(top of the hierarchy) the BorrowingLimit has to be 0.
    //
    // LendingLimit specifies how much ClusterQueues that are NOT under this Cohort
    // can borrow from the ClusterQueues/Cohorts that are under this Cohort.
    // If any of the Limits is not specified it means that there is no limit
    // and ClusterQueues can borrow/lend as much as they want/have.
    // 
    // +listType=atomic
    // +kubebuilder:validation:MaxItems=16
    ResourceGroups []ResourceGroup `json:"resourceGroups,omitempty"`
}

> type CohortStatus struct {

Reviewer 4: Do we need a filed to show the resource borrowed from others?

Author: This may be problematic as with multi-level hierarchy, a single admission may trigger status updates on many levels, increasing the need for etcd write throughput, which remains one of the biggest Kueue bottlenecks. 

I would consider doing the detailed stats for Cohorts via dedicated api, something similar to: https://github.com/kubernetes-sigs/kueue/blob/main/apis/visibility/v1alpha1/types.go

Reviewer 3: Please leave a comment about this in the doc.

Author: Done.

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
      - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit Tests](#unit-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

Introduce Cohort top-level object to allow setting multi-level quota 
hierarchy, with advanced borrowing, lending mechanisms.

## Motivation

The current 2-level hierarchy (ClusterQueues and Cohorts) is not expressive
enough to handle complex use cases of large organizations with tree-like
team and quota/budget structures.

### Goals

* Create a multi-level hierarchy for advanced quota management. 
* Be compatible with the existing ClusterQueue API and mechanics.
* Allow setting constraints about borrowing and lending at all levels.

> * Provide quota for groups of Queues.

Reviewer 4: do we have any component to check if the quota tree is valid?

Author: What do you mean by that? Any non-negative quota is, in theory, valid...

Reviewer 3: That's an implementation detail that is discussed in the risks section.
